src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/StrandArtifact.java

```diff
@@ -0,0 +1,178 @@
+
+        double likelihoodForArtifactFwd = 0.0;
+        double likelihoodForArtifactRev = 0.0;
+        double[] forwardLikelihoodsOfAlleleFractions = new double[numIntegPointsForAlleleFraction];
+        double[] reverseLikelihoodsOfAlleleFractions = new double[numIntegPointsForAlleleFraction];
+
+
+        for (int i = 0; i < numIntegPointsForAlleleFraction; i++) {
+            final double f = gaussIntegrationAbscissasForAlleleFraction.get(i);
+            for (int j = 0; j < numIntegPointsForEpsilon; j++) {
+                final double epsilon = gaussIntegrationAbscissasForEpsilon.get(j);
+                final double integrandArtifactFwd = getIntegrandGivenArtifact(f, epsilon, numReadsForward, numReadsReverse, numAltReadsForward, numAltReadsReverse);
+                final double integrandArtifactRev = getIntegrandGivenArtifact(f, epsilon, numReadsReverse, numReadsForward, numAltReadsReverse, numAltReadsForward);
+
+                //
+                likelihoodForArtifactFwd += gaussIntegrationWeightsForAlleleFraction.get(i) * gaussIntegrationWeightsForEpsilon.get(j) * integrandArtifactFwd;
+                likelihoodForArtifactRev += gaussIntegrationWeightsForAlleleFraction.get(i) * gaussIntegrationWeightsForEpsilon.get(j) * integrandArtifactRev;
+
+                // for a fixed f, integrate over epsilons. this gives us the likelihood p(x^+, x^- | f, z) for a fixed f, which is proportional to
+                // the posterior probability of p(f | x^+, x^-, z)
+                forwardLikelihoodsOfAlleleFractions[i] += gaussIntegrationWeightsForEpsilon.get(j) * integrandArtifactFwd;
+                reverseLikelihoodsOfAlleleFractions[i] += gaussIntegrationWeightsForEpsilon.get(j) * integrandArtifactRev;
+            }
+        }
+
+        unnormalized_posterior_probabilities[StrandArtifactZ.ART_FWD.ordinal()] = pi[StrandArtifactZ.ART_FWD.ordinal()] * likelihoodForArtifactFwd;
+        unnormalized_posterior_probabilities[StrandArtifactZ.ART_REV.ordinal()] = pi[StrandArtifactZ.ART_REV.ordinal()] * likelihoodForArtifactRev;
+
+
+        // find the maximum a posteriori estimate for allele fraction given strand artifact
+        final int maxAlleleFractionIndexFwd = MathUtils.maxElementIndex(forwardLikelihoodsOfAlleleFractions);
+        final int maxAlleleFractionIndexRev = MathUtils.maxElementIndex(reverseLikelihoodsOfAlleleFractions);
+        maximum_a_posteriori_allele_fraction_estimates[StrandArtifactZ.ART_FWD.ordinal()] = gaussIntegrationAbscissasForAlleleFraction.get(maxAlleleFractionIndexFwd);
+        maximum_a_posteriori_allele_fraction_estimates[StrandArtifactZ.ART_REV.ordinal()] = gaussIntegrationAbscissasForAlleleFraction.get(maxAlleleFractionIndexRev);
+
+        // Compute the posterior probability of NO_ARTIFACT; evaluate a single integral over the allele fraction
+        // TODO: write a dot product function and add to MathUtils
```

Comment:
Fo that dot product, you could do `dotProduct(double[] a, double[] b) = MathUtils.sum(MathArrays.ebeMultiply(a,b))` or you could use `RealVector(a).dot(RealVector(b))`.